### PR TITLE
add(api/graphql): query an episode by guid

### LINKED ISSUE #155: GraphQL: query episodes by guid
requested by @alexander-heimbuch for „Preview“

### PR DESCRIPTION
**Pull request by eteubert**
fixes #155

### CODE REVIEW

File: lib/radiator_web/graphql/admin/resolvers/editor.ex [hunk 1]
```diff
@@ -221,6 +228,16 @@ defmodule RadiatorWeb.GraphQL.Admin.Resolvers.Editor do
     end
   end
 
+  def find_episode(_parent, %{guid: guid}, %{context: %{current_user: user}}) do
+    try do
+      with_episode user, guid, :get_episode_by_guid do
+        fn episode -> {:ok, episode} end
+      end
+    rescue
+      Ecto.MultipleResultsError -> {:error, "There is more than one episode with given guid"}
```
Comment (monkeydom) [inline lib/radiator_web/graphql/admin/resolvers/editor.ex]: See above. I'd rather narrow the list down to the ones the user has access to first. And then return. Also not sure if alex didn't want a public api here instead.

File: lib/radiator_web/graphql/schema.ex [hunk 2]
```diff
@@ -66,7 +66,8 @@ defmodule RadiatorWeb.GraphQL.Schema do
 
     @desc "Get one published episode"
     field :published_episode, :published_episode do
-      arg :id, non_null(:id)
+      arg :id, :id
+      arg :guid, :string
```
Comment (monkeydom) [inline lib/radiator_web/graphql/schema.ex]: I think I would prefer a more open endend query api, that also has guid as optional query element and returns a list instead.

File: test/radiator_web/graphql/admin/schema/query/episodes_test.exs [hunk 3]
```diff
@@ -97,6 +97,51 @@ defmodule RadiatorWeb.GraphQL.Admin.Schema.Query.EpisodesTest do
...
+               "episode" => %{
+                 "id" => Integer.to_string(episode.id),
+                 "title" => episode.title,
+                 "guid" => "guid-foo"
+               }
+             }
+           }
+  end
+
+  test "episode by guid returns an episode when non existent", %{conn: conn} do
```
Comment (monkeydom) [inline test/radiator_web/graphql/admin/schema/query/episodes_test.exs]: shouldn't the test names be positive? E.g. `returns not found when non-existent` (and apparently, non-existent is hyphenated according to the macOS dictionary)

File: test/radiator_web/graphql/admin/schema/query/episodes_test.exs [hunk 4]
```diff
@@ -97,6 +97,51 @@ defmodule RadiatorWeb.GraphQL.Admin.Schema.Query.EpisodesTest do
...
+    conn = get conn, "/api/graphql", query: @single_guid_query, variables: %{"guid" => "guid-foo"}
+
+    assert %{"errors" => [%{"message" => message}]} = json_response(conn, 200)
+    assert message =~ "not found"
+  end
+
+  test "episode by guid returns an episode when there is more than one episode with that guid", %{
```
Comment (monkeydom) [inline test/radiator_web/graphql/admin/schema/query/episodes_test.exs]: dito above.

File: lib/radiator_web/graphql/admin/resolvers/editor.ex [hunk 5]
```diff
@@ -72,7 +72,14 @@ defmodule RadiatorWeb.GraphQL.Admin.Resolvers.Editor do
   """
   defmacro with_episode(user, episode_id, do: block) do
     quote do
-      case Editor.get_episode(unquote(user), unquote(episode_id)) do
+      with_episode(unquote(user), unquote(episode_id), :get_episode, do: unquote(block))
+    end
+  end
+
+  defmacro with_episode(user, episode_id, method, do: block) do
```
[the remark this answers is not in the document]
Comment (eteubert) [reply lib/radiator_web/graphql/admin/resolvers/editor.ex]: Renamed it and added an example to the macro doc. Maybe we can revert this whole addition if we go with guid expecting more than one result.

**Comment by eteubert** [comment]
@alexander-heimbuch can you comment on how you expect to use this? Public or Authed API? What would you expect as a result if there are multiple episodes with the same guid in the system?

**Comment by eteubert** [comment]
@monkeydom Since we can't guarantee guids being what their name suggests, globally unique, your proposal to return a list sounds good to me. I would then move the `guid` arg from `:episode` to `:episodes` field (which apparently doesn't even exist yet) and use it as a filter argument there.

**Comment by monkeydom** [comment]
> @monkeydom Since we can't guarantee guids being what their name suggests, globally unique, your proposal to return a list sounds good to me. I would then move the `guid` arg from `:episode` to `:episodes` field (which apparently doesn't even exist yet) and use it as a filter argument there.

Yes, i'd like to see that. Top level query, `publishedEpisodes` and `episodes` with some query args.

**Comment by alexander-heimbuch** [comment]
A `guid` in my specific case is useful if it is a unique identifier. Otherwise I would simply stick to the `id`. Still a more opened query API would very useful for us. For example searching the title/description and also sorting the results.

**Comment by monkeydom** [comment]
Let's shelf this for now.

**Pull request closed without merging**